codex/stores/maintenance.nim

```diff
@@ -10,92 +10,323 @@
+
+        if currDatasetMd.leavesCount != leavesCount:
+          raise newException(CatchableError, "DatasetMetadata for treeCid " & $treeCid & " is already stored with leavesCount " &
+            $currDatasetMd.leavesCount & ", cannot override it with leavesCount " & $leavesCount)
+
+        md.leavesCount = leavesCount
+        md.manifestsCids = (currDatasetMd.manifestsCids & manifestsCids).deduplicate
+        md.checkpoint = Checkpoint(progress: 0, timestamp: 0)
+      else:
+        md.expiry = expiry
+        md.leavesCount = leavesCount
+        md.manifestsCids = manifestsCids
+        md.checkpoint = Checkpoint(progress: 0, timestamp: 0)
+
+      md.some
+  )
+
+proc trackExpiry*(
+  self: DatasetMaintainer,
+  treeCid: Cid,
+  leavesCount: Natural,
+  manifestsCids: seq[Cid] = @[]
+): Future[?!void] {.async.} =
+  await self.trackExpiry(treeCid, leavesCount, self.clock.now + self.defaultExpiry.seconds, manifestsCids)
+
+proc trackExpiry*(
+  self: DatasetMaintainer,
+  cid: Cid,
+  manifestsCids: seq[Cid] = @[]
+): Future[?!void] {.async.} =
+  await self.trackExpiry(cid, 0, self.clock.now + self.defaultExpiry.seconds, manifestsCids)
+
+proc ensureExpiry*(
+  self: DatasetMaintainer,
+  treeCid: Cid,
+  minExpiry: SecondsSince1970): Future[?!void] {.async.} =
+  ## Sets the dataset expiry to a max of two values: current expiry and `minExpiry`,
+  ## if a dataset for given `treeCid` is not currently tracked a CatchableError is thrown
+  ##
+
+  trace "Updating a dataset expiry", treeCid, minExpiry
+
+  without key =? createDatasetMetadataKey(treeCid), err:
+    return failure(err)
+
+  await modify[DatasetMetadata](self.metaDs, key,
+    proc (maybeCurrDatasetMd: ?DatasetMetadata): Future[?DatasetMetadata] {.async.} =
+      if currDatasetMd =? maybeCurrDatasetMd:
+        let datasetMd = DatasetMetadata(
+          expiry: max(currDatasetMd.expiry, minExpiry),
+          leavesCount: currDatasetMd.leavesCount,
+          manifestsCids: currDatasetMd.manifestsCids,
+          checkpoint: currDatasetMd.checkpoint
+        )
+        return datasetMd.some
+      else:
+        raise newException(CatchableError, "DatasetMetadata for treeCid " & $treeCid & " not found")
+  )
+
+
+proc recordCheckpoint*(self: DatasetMaintainer, treeCid: Cid, datasetMd: DatasetMetadata): Future[?!void] {.async.} =
+  # Saves progress or deletes dataset metadata if progress > leavesCount
+  #
+
+  without key =? createDatasetMetadataKey(treeCid), err:
+    return failure(err)
 
-proc runBlockCheck(self: BlockMaintainer): Future[void] {.async.} =
-  let expirations = await self.repoStore.getBlockExpirations(
-    maxNumber = self.numberOfBlocksPerInterval,
-    offset = self.offset
+  await self.metaDs.modify(key,
+    proc (maybeCurrDatasetMd: ?DatasetMetadata): Future[?DatasetMetadata] {.async.} =
+      if currDatasetMd =? maybeCurrDatasetMd:
+        if currDatasetMd.expiry != datasetMd.expiry or currDatasetMd.manifestsCids != datasetMd.manifestsCids:
```

Comment:
>Yikes, can someone re-set the expiry while I'm already garbage-collecting a dataset? Won't that cause data corruption? I suppose you will need some sort of dataset-level locking so that you can't be deleting the blocks in a dataset and storing its blocks at the same time?

A Dataset could be downloaded again while the maintanance has started but hasn't finished yet. It's not an error scenario and I think we should handle it without locking. #864 is indeed an error scenario. If you want I can fix it.